runtime/doc/lua.txt

```diff
@@ -85,6 +85,14 @@ https://www.lua.org/doc/cacm2018.pdf
 - Stackful coroutines enable cooperative multithreading, generators, and
   versatile control for both Lua and its host (Nvim).
 
+                                                          *lua-error-handling*
```

Comment:
todo: move this to the ["lua plugin guide"](https://github.com/neovim/neovim/pull/29073)?

Reply:
I think it's better placed in `develop`, as it describes the pattern _we_ should use in _our_ Lua code. And we should be more discerning:

1. For operations that are expected not to succeed in typical situations ("do if want"), use `nil` for failure, coupled with `assert` (with error message!) if needed.
2. Only throw errors if _users_ need to be informed about failure; otherwise prefer `vim.notify` (and think twice whether it needs to be `ERROR` or `WARN`, or whether it could be `DEBUG` or lower -- we should probably make that configurable).
3. For operations that are expected to work outside of failures beyond our control, use `error` (with message!)
4. For nested calls where the error message requires low-level context but should be triggered at the top-level context (since that is the user-facing interface), prefer `error` and `pcall` over passing error messages as strings through multiple callers.

(`pcall` absolutely has a place, it just shouldn't be used liberally to suppress "printf style" debug messages.)

Reply:
> better placed in `develop`, as it describes the pattern _we_ should use in _our_ Lua code

will do. but lua-guide is also intended for giving hints about best practices, since Lua has no canonical reference for certain practices, and the point of lua-guide is to provide that.

Reply:
I have to disagree; the audience of `lua-guide` is (new) users, not plugin writers or Neovim devs. Adding technical information like that actually reduces its value as an on-boarding guide.

Reply:
> the audience of `lua-guide` is (new) users, not plugin writers or Neovim dev

oh i meant the `lua-plugin` guide, which hasn't been merged yet.

Reply:
I'm keeping `lua-error-handling` in lua.txt in lua-concepts, because it's a fundamental concept. lua.txt is not intended nor restricted to "high level" advice.

I also added `dev-error-patterns` to develop.txt which talks about all available patterns, not only the "result or message" pattern.